Libs/MRML/Core/vtkMRMLBitStreamVolumeNode.cxx

```diff
@@ -0,0 +1,247 @@
+#include "vtkMRMLBitStreamVolumeNode.h"
+
+// MRML includes
+#include "vtkMRMLScene.h"
+#include "vtkMRMLNRRDStorageNode.h"
+
+// VTK includes
+#include <vtkCollection.h>
+#include <vtkMatrix4x4.h>
+#include <vtkNew.h>
+#include <vtkObjectFactory.h>
+#include <vtkXMLUtilities.h>
+
+//----------------------------------------------------------------------------
+vtkMRMLNodeNewMacro(vtkMRMLBitStreamVolumeNode);
+
+//----------------------------------------------------------------------------
+// vtkMRMLBitStreamVolumeNode methods
+
+//-----------------------------------------------------------------------------
+vtkMRMLBitStreamVolumeNode::vtkMRMLBitStreamVolumeNode()
+{
+  this->CompressionDevice = NULL;
+  this->FrameUpdated = false;
+  this->KeyFrameReceived = false;
+  this->KeyFrameDecoded = false;
+  this->KeyFrameUpdated = false;
+  this->FrameMessageValid = false;
+  this->KeyFrameMessage = "";
+  this->FrameMessage = "";
+}
+
+//-----------------------------------------------------------------------------
+vtkMRMLBitStreamVolumeNode::~vtkMRMLBitStreamVolumeNode()
+{
+  if (this->CompressionDevice != NULL)
+    {
+    this->CompressionDevice->Delete();
+    }
+}
+
+//---------------------------------------------------------------------------
+int vtkMRMLBitStreamVolumeNode::ObserveOutsideCompressionDevice(vtkMRMLCompressionDeviceNode* device)
+{
+  if (device)
+    {
+    device->AddObserver(device->GetDeviceContentModifiedEvent(), this, &vtkMRMLBitStreamVolumeNode::ProcessDeviceModifiedEvents);
+    //------
+    if (device->GetContent() != NULL)
+      {
+      std::string bitStream = std::string(device->GetContent()->frameMessage);
+      this->SetFrameMessage(bitStream);
+      if (device->GetContent()->image != NULL)
+        {
+        this->SetAndObserveImageData(device->GetContent()->image);
+        }
+      }
+    this->CompressionDevice = device; // should the interal video device point to the external video device?
+    //-------
+    return 1;
+    }
+  return 0;
+}
+
+void vtkMRMLBitStreamVolumeNode::CopySrcStringToDestString(const std::string& srcString, std::string& destString)
+{
+  destString.resize(srcString.length());
+  char * array = new char[srcString.length()];
+  memcpy(array, srcString.c_str(), srcString.length());
+  destString.assign(array, srcString.length());
+}
+
+void vtkMRMLBitStreamVolumeNode::SetFrameMessage(std::string& buffer)
+{
+  this->CopySrcStringToDestString(buffer, this->FrameMessage);
+  this->FrameMessageValid = true;
```

Comment:
It would be better to not have a separate flag for validity (it may be inconsistent with FrameMessage content), instead, we could just set FrameMessage to NULL or empty when it is not valid.